src/toil/fileStore.py

```diff
@@ -64,10 +70,337 @@ def __init__(self, jobStore, jobWrapper, localTempDir, inputBlockFn):
+        (clean up their presence ont he node, and run any registered defer functions)
+
+        :param nodeInfo: Information regarding the node required for identifying dead jobs.
+        :param bool batchSystemShutdown: Is the batch system in the process of shutting down?
+        :returns: None
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def _registerDeferredFunction(self, callable, *args, **kwargs):
+        """
+        This is the backend function to register a function that will be run after the completion of
+        the current job on the same node
+
+        :param function callable: The function to be run after this job.
+        :param list args: The arguments to the function
+        :param dict kwargs: The keyword arguments to the function
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def _runDeferredFunctions(deferredFunctions):
+        """
+        This function will run all the registered deferred functions for the job.
+
+        :param list deferredFunctions: A list of DeferredFunctions to run the cache
+        :returns: list of failed functions
+        :rtype: list
+        """
+        failures = []
+        for deferredFunction in deferredFunctions:
+            # TODO: better representation of the callable in the output string
+            try:
+                logger.debug('Running deferred function "%s".', deferredFunction.callable.__name__)
+                deferredFunction.callable(*deferredFunction.args, **deferredFunction.kwargs)
+            except:
+                # This has to be generic because we don't know what is in the function
+                failures.append(deferredFunction.callable.__name__)
+                logger.exception('Deferred function "%s" failed.',
+                                 deferredFunction.callable.__name__)
+        return failures
+
+    # Functions related to logging
+    def logToMaster(self, text, level=logging.INFO):
+        """
+        Send a logging message to the leader. The message will also be \
+        logged by the worker at the same level.
+
+        :param text: The string to log.
+        :param int level: The logging level.
+        """
+        logger.log(level=level, msg=("LOG-TO-MASTER: " + text))
+        self.loggingMessages.append(dict(text=text, level=level))
+
+    # Functions run after the completion of the job.
+    @abstractmethod
+    def _updateJobWhenDone(self):
+        """
+        Update the status of the job on the disk.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def _blockFn(self):
+        """
+        Blocks while _updateJobWhenDone is running. This function is called by this job's
+        successor to ensure that it does not begin modifying the job store until after this job has
+        finished doing so.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
```

Comment:
I wouldn't explicitly require this method. It's an implementation detail that subclasses 

The method that you'll introduce for #1240 should be abstract, though.